HSD8-1025 CSV Importer upload form

pookmish opened the pull request:
# READY FOR REVIEW

# Summary
- Provide a UI form to upload a CSV file for migration entities using the `migrate_source_csv` plugin
- Use config overrides to set the path to the csv that was uploaded.

# Need Review By (Date)
- 4/8

# Urgency
- medium

# Steps to Test
1. Checkout this branch and add `migrate_source_csv` module `composer require drupal/migrate_source_csv; drush en migrate_source_csv`
2. Create a simple csv importer. Heres one you can use:
```
langcode: en
status: true
dependencies: {  }
id: csv_news
class: null
field_plugin_method: null
cck_plugin_method: null
migration_tags: {  }
migration_group: default
label: 'CSV News'
source:
  plugin: csv
  path: /tmp/foo.csv
  constants:
    type: stanford_news
  fields:
    -
      name: id
      label: 'ID *Required'
      selector: id
    -
      name: title
      label: 'Title *Required'
      selector: title
  ids:
    - id
process:
  type: constants/type
  su_news_headline: title
destination:
  plugin: 'entity:node'
migration_dependencies:
  required: {  }
```
4. Navigate to the csv uploader
   - If you are using the example above the path is `/admin/structure/migrate/manage/default/migrations/csv_news/csv-upload`
5. download the template file in help text and populate it with some data
6. upload the file and run the importer using drush or the UI `drush mim csv_news` or `/admin/config/importers`
7. verify your content was created.

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

pookmish commented:
@imonroe this PR has been sitting for 2 weeks. H&S is looking to use it soon.

imonroe commented:
While testing: 
- I uploaded a CSV file successfully
- I went into `/admin/structure/migrate/manage/default/migrations/csv_news/csv-upload`
- I opened up `Previously Uploaded Files`
- I checked `Forget previously imported content`
- I get an error: `Drupal\Component\Plugin\Exception\PluginNotFoundException: The "Array" entity type does not exist.`  I think the culprit is `Drupal\stanford_migrate\Form\StanfordMigrateCsvImportForm->save(Array, Object)`

imonroe commented:
My CSV to test with looks like: 
```
"id (ID *Required)","title (Title *Required)"
"103","Title 103"
"104","Title 104"
"105","Title 105"
```

The importer runs, but the titles do not appear to get mapped correctly: 
![image](https://user-images.githubusercontent.com/1263892/115574048-853eda80-a27e-11eb-9164-fc3e2302c39b.png)

pookmish commented:
@imonroe I fixed the error message and found out why the news items didn't have a title. 
https://stanfordits.atlassian.net/browse/D8CORE-2898
It's the way the news node is configured. You can see it at `/admin/structure/types/manage/stanford_news/auto-label`
I've updated the steps to test with the change of `title: title` to `su_news_headline: title`. That should do the trick

imonroe commented on composer.json:
```diff
@@ -19,6 +19,9 @@
         "drupal/ultimate_cron": "^2.0@alpha"
     },
+    "require-dev": {
+        "drupal/migrate_source_csv": "^3.4"
```
Should this be moved to the regular `require` object instead of `require-dev`?

pookmish replied:
It's not a hard requirement. the routes and everything allow it to be an optional dependency.

imonroe commented on tests/src/Kernel/Config/MigrationConfigOverridesTest.php:
```diff
@@ -0,0 +1,63 @@
+      ->get('source');
+    $this->assertArrayNotHasKey('path', $source);
+
+    \Drupal::service('file_system')->copy($this->root . '/core/misc/druplicon.png', 'public://example.jpg');
```
Are you testing with a jpg rather than a csv just for convenience, to make sure the path works?

pookmish replied:
Yeah since we don't really want to test the import since every migration is different.

imonroe reviewed (approved):
This now works as described, and it passes the (very thorough) tests, so I'm good with it.  This is a complex piece of work, nice job.
